### ISSUE TITLE
No meaningful inference when integrating PickNApple checkpoint with SONIC run_sim_loop (MuJoCo)

### ISSUE THREAD

**Issue by yangjunwon1309** [NVIDIA/Isaac-GR00T#661]
**Summary**
Attempting to run the LMPnPAppleToPlateDC_G1 task by integrating a PickNApple checkpoint with the SONIC run_sim_loop MuJoCo simulation environment results in no meaningful inference. I made the custom bridge to connect the published image in SONIC mujoco sim and G1 state from DDS. the bridge file is attached below and it correctly sent the data to GR00T. But The GR00T fails to produce any valid robot behavior yet only standing.

[groot_sonic_bridge.py](https://github.com/user-attachments/files/27366363/groot_sonic_bridge.py)

The table, apple, and plate were added to the original mujoco sim of SONIC as below egocentric image sent via the image publisher.

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/0ff56df3-f550-4675-a895-4069d4fb49c1" />

**Environment**
    Robot: Unitree G1
    Simulation: MuJoCo via SONIC (gear_sonic/scripts/run_sim_loop.py)

**Steps to Reproduce**

### Terminal 1 — SONIC MuJoCo Simulator

```bash
cd ~/GR00T-WholeBodyControl
source .venv_sim/bin/activate
bash start_sonic_sim.sh
```
---

### Terminal 2 — SONIC C++ Deploy

```bash
# Step 1: container 접속 (호스트에서)
cd ~/GR00T-WholeBodyControl
docker run --rm -it \
  --runtime=nvidia --gpus all \
  --privileged --net=host \
  -e NVIDIA_DISABLE_REQUIRE=true \
  -v ~/GR00T-WholeBodyControl:/workspace \
  -v ~/trt_clean:/opt/trt_clean \
  -v /usr/lib/aarch64-linux-gnu/nvidia:/host_nvidia_libs \
  -w /workspace/gear_sonic_deploy \
  g1-deploy-dev:latest /bin/bash
```

```bash
bash start_sonic_deploy.sh
```
---

### Terminal 3 — GR00T Policy Server

```bash
# Step 1
cd ~/Isaac-GR00T
docker run -it --rm --runtime nvidia --gpus all \
  --ipc=host --network host \
  --ulimit memlock=-1 --ulimit stack=67108864 \
  -v ~/Isaac-GR00T:/workspace/repo \
  -v ~/.cache/huggingface:/root/.cache/huggingface \
  -w /workspace/repo \
  gr00t-thor bash
```

```bash
# Step 2
bash /workspace/repo/start_groot_server.sh
```

---

### Terminal 4 — GR00T ↔ SONIC Bridge
```bash
cd ~/Isaac-GR00T
docker run -it --rm --runtime nvidia --gpus all \
  --ipc=host --network host \
  --ulimit memlock=-1 --ulimit stack=67108864 \
  -v ~/Isaac-GR00T:/workspace/repo \
  -v ~/.cache/huggingface:/root/.cache/huggingface \
  -w /workspace/repo \
  gr00t-thor bash
```

The bridge sent the data correctly to GR00T as below
ENDPOINT: get_action
  video.ego_view: shape=(1, 1, 224, 224, 3) dtype=uint8 min=6.000 max=255.000
  
  annotation.human.task_description: ['pick up the apple and place it on the plate']
  
  state.left_leg: shape=(1, 1, 6) dtype=float32 min=-0.150 max=0.107
  
  state.right_leg: shape=(1, 1, 6) dtype=float32 min=-0.456 max=0.211
  
  state.waist: shape=(1, 1, 3) dtype=float32 min=-0.162 max=-0.036
  
  state.left_arm: shape=(1, 1, 7) dtype=float32 min=-0.186 max=0.084
  
  state.right_arm: shape=(1, 1, 7) dtype=float32 min=-0.132 max=0.112
  
  state.left_hand: shape=(1, 1, 7) dtype=float32 min=-0.003 max=-0.000
  
  state.right_hand: shape=(1, 1, 7) dtype=float32 min=-0.000 max=0.063
──
 actions:

  action.left_arm: shape=(1, 30, 7) min=-0.204 max=0.011

  action.right_arm: shape=(1, 30, 7) min=-0.188 max=0.197

  action.left_hand: shape=(1, 30, 7) min=0.000 max=0.000

  action.right_hand: shape=(1, 30, 7) min=-0.029 max=0.064

  action.waist: shape=(1, 30, 3) min=-0.057 max=0.070

  action.base_height_command: shape=(1, 30, 1) min=0.740 max=0.740

  action.navigate_command: shape=(1, 30, 3) min=-0.016 max=0.125

**Comment by rylinjames** [comment]
most likely an obs-action alignment issue at the bridge layer. a few things to check in order:

first, confirm the action chunk is actually being sent to the robot and not silently dropped. add a print right after the policy call to log raw action output before it hits your bridge script. if you're seeing near-zero values there, the issue is in the policy input (normalization or obs format). if the actions look reasonable but the robot doesn't move, the bridge is dropping or scaling them wrong.

second, check your observation normalization stats match what the model was trained with. if you're feeding raw joint positions without normalizing to the model's expected range, the policy sees out-of-distribution input and outputs near-zero actions as a safe fallback.

we built reflex-vla (github.com/FastCrest/reflex-vla) specifically around this class of deployment problem. the `reflex inspect replay` command lets you step through obs-action pairs to catch exactly this kind of mismatch before it hits hardware. might be worth a look.

what does your raw action output look like before the bridge?

**Comment by yangjunwon1309** [comment]
Thanks for your comments!

First, the action chunk from GR00T and inputs for policy (proprioceptive state & ego centric view) are as below.
 * proprioceptive state is received via DDS
 * ego centric view was set as self.camera_configs = {
                "ego_view": {"height": 400, "width": 400, "mjcf_name": "head_camera"},
            }  from basesim.py as given configuration.
 * ego centric view is sent as rgb configuration as image below
-------------------------------------------------------------------------------------------------
DDS init OK
body_q (first 6): [-2.750e-01  6.000e-04 -5.000e-04  6.544e-01 -3.579e-01 -0.000e+00]
body_q nonzero: 29

Calling GR00T get_action...
Inference: 267.9ms

=== RAW ACTION ===
  action.left_arm: mean=0.0653 std=0.2471 min=-0.1458 max=0.6551
    t=0: [ 0.0785 -0.1038  0.0217 -0.1195  0.044  -0.0951  0.6334]
  action.right_arm: mean=0.0943 std=0.2252 min=-0.1051 max=0.6469
    t=0: [ 0.6202 -0.0099  0.0533 -0.0502  0.0258 -0.0524  0.024 ]
  action.left_hand: mean=0.0000 std=0.0000 min=0.0000 max=0.0000
    t=0: [0. 0. 0. 0. 0. 0. 0.]
  action.right_hand: mean=0.0060 std=0.0224 min=-0.0355 max=0.0615
    t=0: [ 0.0205  0.0117  0.0059  0.0293 -0.0176 -0.0137 -0.0191]
  action.waist: mean=-0.0174 std=0.0291 min=-0.0891 max=0.0236
    t=0: [-0.0153 -0.0891  0.0109 -0.0204 -0.0869  0.0109 -0.0102]
  action.base_height_command: mean=0.7400 std=0.0000 min=0.7400 max=0.7400
    t=0: [0.74 0.74 0.74 0.74 0.74 0.74 0.74]
  action.navigate_command: mean=0.0595 std=0.0915 min=-0.1230 max=0.2363
    t=0: [0.022  0.0137 0.0498 0.006  0.0117 0.0364 0.0063]
----------------------------------------------------------------------------------------------

<img width="400" height="400" alt="Image" src="https://github.com/user-attachments/assets/abef7c29-6b2d-4708-8367-bce605ff1a14" />

**Comment by yangjunwon1309** [comment]
is this problem can be on the vlm encoder? or other configurations?

=== MODEL CONFIG ===
  action_horizon: 50
  add_pos_embed: True
  apply_sincos_state_encoding: True
  architectures: ['Gr00tN1d6']
  attn_dropout: 0.2
  attn_implementation: None
  backbone_embedding_dim: 2048
  backbone_model_type: eagle
  backbone_trainable_params_fp32: True
  collator_overwrite_image_inputs: False
  crop_fraction: 0.95
  eagle_collator: True
  formalize_language: True
  gemma_collator: False
  hidden_size: 1024
  image_crop_size: None
  image_target_size: None
  input_embedding_dim: 1536
  load_bf16: True
  max_action_dim: 128
  max_num_embodiments: 32
  max_seq_len: 1024
  max_state_dim: 128
  model_dtype: bfloat16
  model_name: nvidia/Eagle-Block2A-2B-v2
  model_type: Gr00tN1d6
  noise_beta_alpha: 1.5
  noise_beta_beta: 1.0
  noise_s: 0.999
  num_inference_timesteps: 4
  num_timestep_buckets: 1000
  random_rotation_angle: None
  reproject_vision: False
  select_layer: 16
  shortest_image_edge: 256
  state_dropout_prob: 0.0
  torch_dtype: bfloat16
  transformers_version: 4.51.3
  tune_diffusion_model: True
  tune_llm: False
  tune_projector: True
  tune_top_llm_layers: 4
  tune_visual: False
  tune_vlln: True
  use_albumentations_transforms: True
  use_alternate_vl_dit: True
  use_flash_attention: True
  use_relative_action: True
  use_vlln: True

=== BACKBONE CONFIG ===

=== TEXT CONFIG ===

=== EMBODIMENT ===
{
  "robocasa_panda_omron": 13,
  "gr1": 20,
  "behavior_r1_pro": 24,
  "unitree_g1": 8,
  "oxe_google": 0,
  "oxe_widowx": 1,
  "libero_panda": 2,
  "oxe_droid": 16,
  "new_embodiment": 10
}

=== STATISTICS keys ===
['behavior_r1_pro', 'gr1', 'robocasa_panda_omron', 'unitree_g1']

=== FILES ===
['processor_config.json', 'model-00001-of-00002.safetensors', 'trainer_state.json', 'model.safetensors.index.json', 'README.md', 'config.json', 'experiment_cfg', 'embodiment_id.json', 'statistics.json', 'wandb_config.json', 'model-00002-of-00002.safetensors', '.gitattributes', 'assets', '.cache']

**Comment by yangjunwon1309** [comment]
and the robot is moving forward (keep moving forward) if i turn off the image sending in bridge. why is that happens?

**Comment by yangjunwon1309** [comment]
example cmd by GR00T N1.6 in sonic policy is as below, keep standing and abruptly moving its arm.

Replanning with mode: WALK, target_height: 0.74, target_vel: 0.0688667, movement: [0.31197, 0.950092, 0], facing: [0.31197, 0.950092, 0]
Replanning with mode: IDLE, target_height: 0.74, target_vel: -1, movement: [0, 0, 0], facing: [1, 0, 0]
Replanning with mode: WALK, target_height: 0.74, target_vel: 0.0850507, movement: [0.264089, 0.964498, 0], facing: [0.264089, 0.964498, 0]
Replanning with mode: WALK, target_height: 0.74, target_vel: 0.0616938, movement: [0.191929, 0.981409, 0], facing: [0.191929, 0.981409, 0]
Replanning with mode: WALK, target_height: 0.74, target_vel: 0.0900346, movement: [0.412168, 0.911108, 0], facing: [0.412168, 0.911108, 0]
Replanning with mode: IDLE, target_height: 0.74, target_vel: -1, movement: [0, 0, 0], facing: [1, 0, 0]
Loop timing - LowState age: 16.28ms, Streaming data mean delay: 0ms, Streaming data std delay: 0ms, IMU age: 16.221ms, Obs: 7086us, Policy: 7227us, Obs 2 Motor Command: 14314us, Post processing: 154us, Planner - Gather Input: 57us, Model: 11508us, Convert50Hz: 15us, Total: 11582us | HandCloseRatio: 1
Replanning with mode: WALK, target_height: 0.74, target_vel: 0.0667447, movement: [-0.197523, 0.980298, 0], facing: [-0.197523, 0.980298, 0]
Replanning with mode: IDLE, target_height: 0.74, target_vel: -1, movement: [0, 0, 0], facing: [1, 0, 0]
Loop timing - LowState age: 2.941ms, Streaming data mean delay: 0ms, Streaming data std delay: 0ms, IMU age: 2.917ms, Obs: 656us, Policy: 353us, Obs 2 Motor Command: 1009us, Post processing: 57us, Planner - Gather Input: 67us, Model: 6530us, Convert50Hz: 15us, Total: 6614us | HandCloseRatio: 1
Replanning with mode: WALK, target_height: 0.74, target_vel: 0.0665569, movement: [0.381488, 0.924374, 0], facing: [0.381488, 0.924374, 0]
Replanning with mode: IDLE, target_height: 0.74, target_vel: -1, movement: [0, 0, 0], facing: [1, 0, 0]
Replanning with mode: WALK, target_height: 0.74, target_vel: 0.0640821, movement: [0.137153, 0.99055, 0], facing: [0.137153, 0.99055, 0]
Replanning with mode: IDLE, target_height: 0.74, target_vel: -1, movement: [0, 0, 0], facing: [1, 0, 0]
Replanning with mode: WALK, target_height: 0.74, target_vel: 0.0670992, movement: [-0.363851, 0.931457, 0], facing: [-0.363851, 0.931457, 0]
Replanning with mode: IDLE, target_height: 0.74, target_vel: -1, movement: [0, 0, 0], facing: [1, 0, 0]
Loop timing - LowState age: 1.844ms, Streaming data mean delay: 0ms, Streaming data std delay: 0ms, IMU age: 1.791ms, Obs: 721us, Policy: 435us, Obs 2 Motor Command: 1156us, Post processing: 62us, Planner - Gather Input: 96us, Model: 8725us, Convert50Hz: 15us, Total: 8837us | HandCloseRatio: 1
Replanning with mode: WALK, target_height: 0.74, target_vel: 0.0817751, movement: [0.644871, 0.764291, 0], facing: [0.644871, 0.764291, 0]
Replanning with mode: WALK, target_height: 0.74, target_vel: 0.0761781, movement: [-0.0128195, 0.999918, 0], facing: [-0.0128195, 0.999918, 0]
Replanning with mode: IDLE, target_height: 0.74, target_vel: -1, movement: [0, 0, 0], facing: [1, 0, 0]

**Comment by yangjunwon1309** [comment]
After I changed the checkpoint into the original nvidia checkpoint (https://huggingface.co/nvidia/GR00T-N1.6-G1-PnPAppleToPlate), the robot moved forward to the table and tried to pick up the apple (by gripping motion). But, failed to grip. So, I think the problem was on the overfitted finetuned checkpoint I used, so that it was not worked on the new environment.

**Comment by rylinjames** [comment]
that is a very clean self-diagnose. overfitted checkpoint explains the complete failure on the new environment.

for the remaining grip failure on the original nvidia checkpoint, this is almost always a gripper action range mismatch. GR00T outputs gripper dims typically in [-1, 1], but the Inspire hand controller on G1 often expects [0, 1]. print the raw gripper action output from the policy before your bridge remaps it. if you see negative values or near-zero when the hand should be closing, that is your mismatch.

also double check you are loading the original checkpoint's norm_stats.json, not your finetuned one. the norm stats drift between finetuned and base checkpoints can shift the gripper dim enough that close becomes almost open.

if you want real-time per-dimension action monitoring on the robot, reflex-vla's bridge inspector (github.com/FastCrest/reflex-vla) shows exactly what each joint is doing at the inference boundary. but first verify the gripper range.

**Comment by yangjunwon1309** [comment]
Thanks a lot. But, when I tried to install the reflex-vla and use it, the current version doesn't seems to support the GR00T N1.6 and its familes yet (only pi series, smol VLA, and etc were on the options). And the internal doctor and chatbot services also applied to me that the GR00T is not on the service yet. So, is it possible to run the GR00T on your reflex-vla now?
